pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentCreationMapper.java

```diff
@@ -183,13 +184,24 @@ protected void map(LongWritable key, Text value, Context context) {
       LOGGER.info("Size for segment: {}, uncompressed: {}, compressed: {}", segmentName,
           DataSizeUtils.fromBytes(uncompressedSegmentSize), DataSizeUtils.fromBytes(compressedSegmentSize));
       //move segment to output PinotFS
-      URI outputSegmentTarURI = SegmentGenerationUtils.getRelativeOutputPath(inputDirURI, inputFileURI, outputDirURI)
-          .resolve(segmentTarFileName);
-      LOGGER.info("Copying segment tar file from [{}] to [{}]", localSegmentTarFile, outputSegmentTarURI);
-      outputDirFS.copyFromLocalFile(localSegmentTarFile, outputSegmentTarURI);
+      URI relativeOutputPath = SegmentGenerationUtils.getRelativeOutputPath(inputDirURI, inputFileURI, outputDirURI);
+      URI outputSegmentTarURI = relativeOutputPath.resolve(segmentTarFileName);
+      SegmentGenerationJobUtils.moveLocalTarFileToRemote(localSegmentTarFile, outputSegmentTarURI,
+          _spec.isOverwriteOutput());
 
+
+      String metadataTarFileName = URLEncoder.encode(segmentName + Constants.METADATA_TAR_GZ_FILE_EXT, "UTF-8");
```

Comment:
When `_spec.isOverwriteOutput()` is false and `taskSpec.isCreateMetadataTarGz()` is true, we should not delete `outputMetadataTarURI`. Otherwise, segment tar file will NOT be replaced and metadata tar file will be replaced, which causes inconsistency.

Reply:
> > > Doesn't this introduce inconsistency between the two steps. Assume the following scenario -
> > > 
> > > * Generate one segment tar using StandaloneJobRunner, we also generate the corresponding metadata tar for it
> > > * Replaced that segment tar in the output dir with some other tar file with the same name
> > > * We trigger metadata push which pushes the old metadata tar, however the segment tar file contains different data
> > 
> > 
> > You are 100% correct. I can make the segment creation job to delete the metadata tarball as well if not exists. But if users intend to manually update the segment tarball, then it's pretty hard to detect that.
> 
> Thinking out loud here, but may be we can introduce a crc check and keep segment tar crc in the metadata tar?

The purpose of using metadata push is to avoid the segment download from controller. The CRC is in segment metadata, but we cannot access it without download the segment.

Reply:
I think we will delete the metadata tar gz file once `_spec.isOverwriteOutput()` is true.
Just in the case of user switch on then off the `isCreateMetadataTarGz` flag. Then this code will only generate segment tar gz and delete the metadata tar gz file.

Reply:
Thanks for the udpate!

I feel that there are 3 cases
1. `_spec.isOverwriteOutput()` is true -> we need to delete metadata from output dir when it exists
2. `_spec.isOverwriteOutput()` is false and `taskSpec.isCreateMetadataTarGz` is true, we should try to create metadata targz locally and try to copy it to output dir in case the metadata targz does not exist (due to user switches off then on the `isCreateMetadataTarGz` flag)
3. `_spec.isOverwriteOutput()` is false and `taskSpec.isCreateMetadataTarGz` is false, we should try to delete metadata targz from output dir in case the metadata targz exists (due to user switches on then off the `isCreateMetadataTarGz` flag)

I think we missed case 3, what do you think?